File: go/day23_part1_2016.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"strconv"
	"strings"
)

func main() {
	instructions := readInstructions("input.txt")
	registers := map[string]int{"a": 7, "b": 0, "c": 0, "d": 0} // Initialize register 'a' to 7 (number of eggs)
	executeInstructions(instructions, registers)
	fmt.Println(registers["a"])
}

func readInstructions(filename string) []string {
	file, err := os.Open(filename)
	if err != nil {
		panic(err)
	}
	defer file.Close()

	var instructions []string
	scanner := bufio.NewScanner(file)
	for scanner.Scan() {
		instructions = append(instructions, scanner.Text())
	}
	return instructions
}

func executeInstructions(instructions []string, registers map[string]int) {
	var pc int // Program counter
	for pc < len(instructions) {
		fields := strings.Fields(instructions[pc])
		switch fields[0] {
		case "cpy":
			x := getValue(fields[1], registers)
			if _, exists := registers[fields[2]]; exists {
				registers[fields[2]] = x
			}
		case "inc":
			if _, exists := registers[fields[1]]; exists {
				registers[fields[1]]++
			}
		case "dec":
			if _, exists := registers[fields[1]]; exists {
				registers[fields[1]]--
			}
		case "jnz":
			x := getValue(fields[1], registers)
			if x != 0 {
				pc += getValue(fields[2], registers) - 1
			}
		case "tgl":
			x := getValue(fields[1], registers)
			if tgt := pc + x; tgt >= 0 && tgt < len(instructions) {
				instructions[tgt] = toggleInstruction(instructions[tgt])
			}
		}
		pc++
	}
}

func getValue(s string, registers map[string]int) int {
	if val, err := strconv.Atoi(s); err == nil {
		return val
	}
	return registers[s]
}

func toggleInstruction(instr string) string {
	parts := strings.Fields(instr)
	switch parts[0] {
	case "inc":
		parts[0] = "dec"
	case "dec", "tgl":
		parts[0] = "inc"
	case "jnz":
		parts[0] = "cpy"
	case "cpy":
		parts[0] = "jnz"
	}
	return strings.Join(parts, " ")
}

File: csharp/day23_part1_2016.cs

using System;
using System.Collections.Generic;
using System.IO;

class Program
{
    static void Main()
    {
        string[] instructions = File.ReadAllLines("input.txt");
        Dictionary<string, int> registers = new Dictionary<string, int> { { "a", 7 }, { "b", 0 }, { "c", 0 }, { "d", 0 } };
        ExecuteInstructions(instructions, registers);
        Console.WriteLine(registers["a"]);
    }

    static void ExecuteInstructions(string[] instructions, Dictionary<string, int> registers)
    {
        int pc = 0;
        while (pc < instructions.Length)
        {
            string[] fields = instructions[pc].Split(' ');
            switch (fields[0])
            {
                case "cpy":
                    int x = GetValue(fields[1], registers);
                    if (registers.ContainsKey(fields[2]))
                    {
                        registers[fields[2]] = x;
                    }
                    break;
                case "inc":
                    if (registers.ContainsKey(fields[1]))
                    {
                        registers[fields[1]]++;
                    }
                    break;
                case "dec":
                    if (registers.ContainsKey(fields[1]))
                    {
                        registers[fields[1]]--;
                    }
                    break;
                case "jnz":
                    int val = GetValue(fields[1], registers);
                    if (val != 0)
                    {
                        pc += GetValue(fields[2], registers) - 1;
                    }
                    break;
                case "tgl":
                    int target = pc + GetValue(fields[1], registers);
                    if (target >= 0 && target < instructions.Length)
                    {
                        instructions[target] = ToggleInstruction(instructions[target]);
                    }
                    break;
            }
            pc++;
        }
    }

    static int GetValue(string s, Dictionary<string, int> registers)
    {
        if (int.TryParse(s, out int val))
        {
            return val;
        }
        return registers[s];
    }

    static string ToggleInstruction(string instr)
    {
        string[] parts = instr.Split(' ');
        switch (parts[0])
        {
            case "inc":
                parts[0] = "dec";
                break;
            case "dec":
            case "tgl":
                parts[0] = "inc";
                break;
            case "jnz":
                parts[0] = "cpy";
                break;
            case "cpy":
                parts[0] = "jnz";
                break;
        }
        return string.Join(" ", parts);
    }
}
